add module Data.ObjectMap.ST

gbagan opened the pull request:
Hi,

Building an ObjectMap in a purely functional way can be very expensive since insertions take O(n) time.
So, I added a module to be able to manipulate mutable ObjectMaps via the ST monad.

I have also removed non exported modules in test/Main.purs, created a file test.dhall and removed libraries from spago.dhall that only concern tests.
You can run test via `spago test -x test.dhall`

m-bock commented on test/Main.purs:
```diff
@@ -47,3 +47,18 @@ main = launchAff_ $ runSpec [ consoleReporter ] do
+            _ <- STOM.poke 2 3 m
+            _ <- STOM.poke 3 3 m
+            _ <- STOM.poke 1 1 m
+            _ <- STOM.delete 3 m
```
I think I'd remove the `delete` part from this test, so it just shows that the odd order of "2,3,1" is stable. Like in the other tests.

m-bock commented on test.dhall:
```diff
@@ -0,0 +1,5 @@
+let conf = ./spago.dhall
```
Yes, much better with the test.dhall!

m-bock commented on src/Data/ObjectMap.purs:
```diff
@@ -1,11 +1,12 @@
 module Data.ObjectMap
-  ( ObjectMap
+  ( ObjectMap(..)
```
Ok, I see. You need the constructor for `unsafeFreeze` in the other module. I would prefer to find another solution here. I think hiding the implementation is important.

- I am not so experienced with `ST`, so I ask: do we need `unsafeFreeze`?
- If we need it, can we implement it without the constructor?
- Otherwise I'd go for the option to introduce an `.Internal` module that exposes the constructor and the main module would still hide it.

gbagan replied:
Yes, I understand your point.

`unsafeFreeze` (or `freeze` that does a copy) is the only way to convert a mutable ObjectMap to a Map and thus to "leave" the ST monad.
I don't see how to implement it without the constructor so I guess intoducing an `Internal` module is the best solution.
If you agree, I will do that and update the README.

thought2 reviewed (changes requested):
Great, thanks for the contribution. I'd like to merge it but I left some comments for discussion.
I think we should also update the README with some better news about the insertion performance.

gbagan commented:
I have moved `ObjectMap` definition and related functions to `Data.ObjectMap.Internal`
So, now, the constructor of `ObjectMap` is hidden in `Data.ObjectMap`.
It was necessary to move related functions to avoid class orphans for `Index` and `At`.
I have updated README.
Feel free to rewrite it, I am a poor documentation writer.

thought2 commented:
Great, I added some small fixes, they're in the `pr/1` branch right now. If you rebase or merge them into your branch, I'll merge to main and release the next version of the package to the registry.

Unrelated:
While looking at my code again I saw that in the future I may want to get rid of the `EncodeJson` instance. Opened an issue for this, maybe you have some feedback on this.
